From a11c34ec2986e88d166a51c1d7835ca25be27d16 Mon Sep 17 00:00:00 2001
From: Evan Chen <evanxinchen@google.com>
Date: Tue, 19 Aug 2025 11:18:40 -0700
Subject: [PATCH] Make sure to disassociate all the packages with shared UID

Test: cts
Bug: 427206637
Flag: Flag: EXEMPT bugfix
Change-Id: I751d724cefe415000646d1640c65e74af52a6688
Merged-In: I3cccad4e85e77916cff020eb7087ac2c355e3e23
---
 .../CompanionDeviceManagerService.java        | 28 +++++++++++--------
 1 file changed, 17 insertions(+), 11 deletions(-)

diff --git a/services/companion/java/com/android/server/companion/CompanionDeviceManagerService.java b/services/companion/java/com/android/server/companion/CompanionDeviceManagerService.java
index 194789bcbbb2..4b1bf0e3c14b 100644
--- a/services/companion/java/com/android/server/companion/CompanionDeviceManagerService.java
+++ b/services/companion/java/com/android/server/companion/CompanionDeviceManagerService.java
@@ -166,6 +166,7 @@ public class CompanionDeviceManagerService extends SystemService {
     private final IAppOpsService mAppOpsManager;
     private final PowerWhitelistManager mPowerWhitelistManager;
     private final UserManager mUserManager;
+    private final PackageManager mPackageManager;
     final PackageManagerInternal mPackageManagerInternal;
 
     /**
@@ -219,6 +220,7 @@ public class CompanionDeviceManagerService extends SystemService {
         mAtmInternal = LocalServices.getService(ActivityTaskManagerInternal.class);
         mAmInternal = LocalServices.getService(ActivityManagerInternal.class);
         mPackageManagerInternal = LocalServices.getService(PackageManagerInternal.class);
+        mPackageManager = context.getPackageManager();
         mUserManager = context.getSystemService(UserManager.class);
 
         mUserPersistenceHandler = new PersistUserStateHandler();
@@ -1556,8 +1558,10 @@ public class CompanionDeviceManagerService extends SystemService {
                 return;
             }
 
-            final String packageName = getPackageNameByUid(uid);
-            if (packageName == null) {
+            // A UID can be shared by multiple packages if android:sharedUserId is used.
+            // We must get all packages for the UID to ensure we find the correct one.
+            final String[] packageNames = mPackageManager.getPackagesForUid(uid);
+            if (packageNames == null || packageNames.length == 0) {
                 // Not interested in this uid.
                 return;
             }
@@ -1566,17 +1570,19 @@ public class CompanionDeviceManagerService extends SystemService {
 
             boolean needToPersistStateForUser = false;
 
-            for (AssociationInfo association :
-                    getPendingRoleHolderRemovalAssociationsForUser(userId)) {
-                if (!packageName.equals(association.getPackageName())) continue;
+            for (String packageName : packageNames) {
+                for (AssociationInfo association :
+                        getPendingRoleHolderRemovalAssociationsForUser(userId)) {
+                    if (!packageName.equals(association.getPackageName())) continue;
 
-                if (!maybeRemoveRoleHolderForAssociation(association)) {
-                    // Did not remove the role holder, will have to try again later.
-                    continue;
-                }
+                    if (!maybeRemoveRoleHolderForAssociation(association)) {
+                        // Did not remove the role holder, will have to try again later.
+                        continue;
+                    }
 
-                removeFromPendingRoleHolderRemoval(association);
-                needToPersistStateForUser = true;
+                    removeFromPendingRoleHolderRemoval(association);
+                    needToPersistStateForUser = true;
+                }
             }
 
             if (needToPersistStateForUser) {
-- 
2.52.0.rc1.455.g30608eb744-goog

